subsys/dfu/src/dfu_target_flash.c

```diff
@@ -0,0 +1,168 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+
+/* TODO future improvements:
+ * - settings context to have resume on reboot like mcuboot target
+ * - use flash_area API, to get this PM needs spi flash info - flash_map
+ * - support progressive erase
+ * - ideally re-use flash_img.c in some manner.
+ */
+
+#include <zephyr.h>
+#include <drivers/flash.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(dfu_target_flash, CONFIG_DFU_TARGET_LOG_LEVEL);
+
+#define FLASH_HEADER_MAGIC 0x7fdeb13c
+
+static struct device *flash_dev;
+static const struct flash_pages_layout *layout;
+
+static u8_t *write_buf;
+static size_t write_buf_pos;
+
+static int offset;
+static size_t available_size;
+static int start_address;
+
+static int write_buf_to_flash(void)
+{
+	int rc;
+	int write_address = start_address + offset;
+
+	flash_write_protection_set(flash_dev, false);
+	rc = flash_erase(flash_dev, write_address, layout->pages_size);
+	if (rc != 0) {
+		LOG_ERR("flash erase failed: %d", rc);
+		goto out;
+	}
+
+	rc = flash_write(flash_dev, write_address, write_buf,
+			layout->pages_size);
+	if (rc != 0) {
+		LOG_ERR("flash write failed: %d", rc);
+		goto out;
+	}
+
+out:
+	flash_write_protection_set(flash_dev, true);
+	return rc;
+}
+
+int dfu_target_flash_cfg(const char *dev_name, size_t start, size_t end,
+			 void *buf, size_t buf_len)
+{
+	size_t total_size;
+	size_t layout_size;
+
+	flash_dev = device_get_binding(dev_name);
+	if (!flash_dev) {
+		LOG_ERR("Unable to initialize with configured device");
+		return -EFAULT;
+	}
+
+	const struct flash_driver_api *api = flash_dev->driver_api;
+
+	api->page_layout(flash_dev, &layout, &layout_size);
+
+	/* Only single sized page layout supported */
+	__ASSERT_NO_MSG(layout_size == 1);
+
+	start_address = start;
+	total_size = layout->pages_count * layout->pages_size;
+
+	if (end > total_size ||
```

Comment:
`(end - start) > total_size` ?

Reply:
Then it makes more sense to ensure that `end > start` don't you think?